rest/multipart.go

```diff
@@ -27,30 +27,56 @@ const kMaxInlineAttachmentSize = 200
 // JSON bodies smaller than this won't be GZip-encoded.
 const kMinCompressedJSONSize = 300
 
-// Parses a JSON MIME body, unmarshaling it into "into".
+// ReadJSONFromMIME parses a JSON MIME body, unmarshalling it into "into".
 // Closes the input io.ReadCloser once done.
 func ReadJSONFromMIME(headers http.Header, input io.ReadCloser, into interface{}) error {
-	contentType := headers.Get("Content-Type")
-	if contentType != "" && !strings.HasPrefix(contentType, "application/json") {
-		return base.HTTPErrorf(http.StatusUnsupportedMediaType, "Invalid content type %s", contentType)
+	// Performs the Content-Type validation and Content-Encoding check.
+	input, err := processContentEncoding(headers, input)
+	if err != nil {
+		return err
 	}
 
-	switch headers.Get("Content-Encoding") {
-	case "gzip":
-		var err error
-		if input, err = gzip.NewReader(input); err != nil {
-			return err
+	// Decode the body bytes into target structure.
+	err = base.JSONDecoderWithOptions(input, true, true).Decode(&into)
+	if err != nil {
+		err = base.WrapJSONUnknownFieldErr(err)
+		if errors.Cause(err) == base.ErrUnknownField {
+			err = base.HTTPErrorf(http.StatusBadRequest, "JSON Unknown Field: %s", err.Error())
+		} else {
+			err = base.HTTPErrorf(http.StatusBadRequest, "Bad JSON: %s", err.Error())
 		}
-	case "":
-		break
-	default:
-		return base.HTTPErrorf(http.StatusUnsupportedMediaType, "Unsupported Content-Encoding; use gzip")
+	}
+	_ = input.Close()
+	return err
+}
+
+// ReadSanitizeConfigJSON parses a JSON body, unmarshals it into "DbConfig".
+// Closes the input io.ReadCloser once done. It expands environment variables and sanitizes the
+// input bytes by converting the back quotes into double-quotes, escaping literal backslashes,
+// newlines or double-quotes with backslashes.
+func ReadSanitizeConfigJSON(headers http.Header, input io.ReadCloser, config *DbConfig) error {
+	// Performs the Content-Type validation and Content-Encoding check.
+	input, err := processContentEncoding(headers, input)
+	if err != nil {
+		return err
 	}
 
-	decoder := base.JSONDecoder(input)
-	decoder.DisallowUnknownFields()
-	decoder.UseNumber()
-	err := decoder.Decode(into)
+	// Read body bytes to sanitize the content and substitute environment variables.
+	content, err := ioutil.ReadAll(input)
```

Comment:
If we get an error doing ReadAll, should we still be calling input.Close() before returning?

Reply:
Yeah, this wouldn't be calling input.Close() before returning but we need to close the input before returning. Added changes in the latest commit.